### ISSUE TITLE
Receiving websocket payloads larger than ~4K yields garbled data

### ISSUE THREAD

**Issue by cemerick** [camlworks/dream#214]
Here's a minimal reproduction that lets one send arbitrarily-sized websocket payloads to the server; sizes > ~4K results in garbled (seemingly arbitrary) data:

```ocaml
let rec bus_loop ws =
  let open Lwt.Infix in
  receive ws
  >>= function
  | None     -> close_websocket ws
  | Some msg ->
    let open String in
    let len = Int.min 40 (length msg) in
    log "Message (%d) ending with: %s" (length msg) (sub msg (length msg - len) len);
    bus_loop ws

let () =
  Lwt_main.run
  @@ serve ~interface:"0.0.0.0" ~port:9090
       (router
          [ get "/ws" (fun _req -> websocket bus_loop)
          ; (get "/"
            @@ fun _r ->
            Dream.respond
              {|<html>
           websocket payload size:
           <input type="number" id="file" />
           <button>Send</button>
           <script>
           function send (input) {
             let ws = new WebSocket("ws://localhost:9090/ws");
             ws.addEventListener("open", e => {
               let bt = ws.binaryType;
               let len = parseInt(input.value);
               let data = new Uint8Array(len);
               data.fill(33, 0, len);
               ws.binaryType = 'arraybuffer';
               ws.send(data);
               ws.binaryType = bt;
             });
           }

           document.querySelector("button").addEventListener("click", e => send(document.querySelector("input")));
           </script>
           </html>|}
            )
          ])
```

A sample output log looks like this (note that the messages being sent by the client consist entirely of exclamation points, just to easily see where the bad data starts):

```
29.03.22 03:18:38.261                       Message (4000) ending with: !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
29.03.22 03:18:44.169                       Message (4090) ending with: !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!�9
29.03.22 03:21:34.808                       Message (5000) ending with: ����������������������������������������
29.03.22 03:21:37.178                       Message (5000) ending with: wXwXwXwXwXwXwXwXwXwXwXwXwXwXwXwXwXwXwXwX
```
 Of course maybe there's another perfectly reasonable cause, it sure feels like the websocket facility is successfully handling the first "frame" of data (which is probably 4096 bytes, including some header data, yielding something like 4000 bytes of expected data), but then only increases the extent of the resulting string, without actually copying over the payload of websocket "frame" >= 2.

**Comment by cemerick** [comment]
(BTW, I'll also gladly accept the news that I'm somehow running afoul of some necessary yet missing websocket handling :worried:)

**Comment by cemerick** [comment]
When I attempt to send much larger payloads (say, 20MB), more stranger things happen:

1. a single payload sent (sometimes!) results in multiple messages received, as if the original single payload is not recognized as such
2. the messages that are received have strange sizes, starting with one larger message, and then a pile of much smaller messages. Example output for that case:

```
29.03.22 13:11:51.406                       Message (131099) ending with: c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�
29.03.22 13:11:51.406                       Message (99) ending with: c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�
29.03.22 13:11:51.406                       Message (99) ending with: c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�c=,�
```

...with pages and pages of log indicating that, in this case, tons of 99-byte messages were received.

That larger message often seems to be something close to 131000 bytes, which seems just as suspiciously a round number as the ~4K threshold I noted in my original report. :thinking:

**Comment by aantron** [comment]
> (BTW, I'll also gladly accept the news that I'm somehow running afoul of some necessary yet missing websocket handling 😟)

This is almost certainly a problem in Dream or websocket/af. I unfortunately won't be able to look at this immediately, but I am reading the very helpful information and will try to address this soon.

**Comment by cemerick** [comment]
I fished around in both dream and websocketaf hoping to track the specific cause of this down, but without ultimate success. Some tidbits that might be helpful though:

1. When receiving messages > ~4K, Dream does not receive fragments from the underlying stream when it should; specifically, it sometimes requires 2+ top-level messages being sent before a deluge of fragments come through (with the garbled data as described before).
2. Chrome's websocket messages panel sometimes reports `Received a broken close frame containing a reserved status code.` after a number of too-large messages are sent.

My hunch is that websocketaf is what's misbehaving. If Dream's streams were at fault, then things like large multipart form uploads should be breaking too; and that "broken close frame" report is probably not something that Dream's higher-level websocket support could provoke?

Until this is resolved, my workaround is to just add my own (3kb, yeuch lol) framing to known-large websocket messages. Obviously not ideal, but it seems to work well enough in the interim.

**Comment by aantron** [comment]
I've recently been debugging this and this appears to be an upstream Websocket/af issue, as probably also suggested by https://github.com/anmonteiro/websocketaf/issues/48, because this can be worked around by changing the buffer size in Websocket/af. As far as I can tell, Dream already receives garbled data for fragments after the first, and I don't think Dream is clobbering or can clobber Webscoket/af's internal buffer. So I don't see how Dream can fix the reassembly of these messages.

cc @anmonteiro

**Comment by cemerick** [comment]
Yes, it'd definitely a problem upstream of Dream, and @anmonteiro and I have sporadically gone back and forth on the subject. I think I currently owe him a minimal repro app, which I hope to get to assemble in the next week or so.

**Comment by anmonteiro** [comment]
Commenting here for both this one and https://github.com/aantron/dream/issues/230

The general situation of my websocketaf fork is that it's an experiment to check whether it could work. Unfortunately, and as you've noticed in Dream, it's not meant to be used at all (unlike h2/httpaf, which are running production workloads).

I suppose that the former doesn't matter when it's indeed used in Dream, and we should strive to make the library better.

Unfortunately I don't have any free time to spare right now on websocketaf – there seems to be other interest in using it too from other community folks. I'm happy to review PRs or otherwise talk about a paid engagement to make websocketaf ready for dream.

**Comment by anmonteiro** [comment]
I finally found the root cause of this issue and have fixed it in https://github.com/anmonteiro/httpun-ws/pull/68
